File: sdk/python/pact/revocation.py
"""Revocation support for the Pact protocol.

A delegator can revoke a delegation they previously issued by creating
a signed Revocation object.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Optional

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

from .canonical import canonical_json
from .encoding import decode_base64url, encode_base64url
from .identity import Identity, derive_id


@dataclass
class Revocation:
    """A signed statement revoking a previously issued delegation."""

    type: str = "revocation"
    delegation_id: str = ""
    revoked_by: str = ""
    revoked_at: str = ""
    reason: str = ""
    signature: str = ""

    def to_dict(self) -> dict:
        d: dict = {
            "type": self.type,
            "delegation_id": self.delegation_id,
            "revoked_by": self.revoked_by,
            "revoked_at": self.revoked_at,
        }
        if self.reason:
            d["reason"] = self.reason
        d["signature"] = self.signature
        return d

    @classmethod
    def from_dict(cls, d: dict) -> Revocation:
        return cls(
            type=d.get("type", "revocation"),
            delegation_id=d.get("delegation_id", ""),
            revoked_by=d.get("revoked_by", ""),
            revoked_at=d.get("revoked_at", ""),
            reason=d.get("reason", ""),
            signature=d.get("signature", ""),
        )

    def signing_payload(self) -> bytes:
        """Return the canonical JSON bytes that are signed."""
        d = self.to_dict()
        d["signature"] = ""
        return canonical_json(d)


def new_revocation(
    delegation_id: str,
    revoker: Identity,
    reason: str = "",
) -> Revocation:
    """Create a signed revocation for a delegation.

    Args:
        delegation_id: The ID of the delegation to revoke.
        revoker: The entity who issued the original delegation (must have private key).
        reason: Optional human-readable reason.

    Returns:
        A signed Revocation.

    Raises:
        ValueError: If the revoker has no private key.
    """
    if revoker.private_key is None:
        raise ValueError("Revoker must have a private key to sign")

    rev = Revocation(
        type="revocation",
        delegation_id=delegation_id,
        revoked_by=revoker.id,
        revoked_at=datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        reason=reason,
        signature="",
    )

    # Sign the revocation
    payload = rev.signing_payload()
    sig = revoker.sign(payload)
    rev.signature = encode_base64url(sig)

    return rev


def verify_revocation(revocation: Revocation, revoker_public_key: str) -> bool:
    """Verify that a revocation was properly signed by the revoker.

    Args:
        revocation: The revocation to verify.
        revoker_public_key: Base64url-encoded public key of the revoker.

    Returns:
        True if the revocation is valid, False otherwise.
    """
    try:
        pub_bytes = decode_base64url(revoker_public_key)
        pub_key = Ed25519PublicKey.from_public_bytes(pub_bytes)

        # Verify the ID matches the key
        expected_id = derive_id(pub_bytes)
        if expected_id != revocation.revoked_by:
            return False

        # Decode signature
        sig_bytes = decode_base64url(revocation.signature)

        # Reconstruct payload
        payload = revocation.signing_payload()

        # Verify
        pub_key.verify(sig_bytes, payload)
        return True
    except Exception:
        return False


class MemoryRevocationStore:
    """In-memory revocation store for checking if delegations are revoked.

    Suitable for single-process providers. For distributed systems,
    implement your own store backed by Redis, PostgreSQL, etc.
    """

    def __init__(self) -> None:
        self._revoked: dict[str, Revocation] = {}

    def revoke(self, revocation: Revocation) -> None:
        """Add a revocation to the store."""
        self._revoked[revocation.delegation_id] = revocation

    def is_revoked(self, delegation_id: str) -> bool:
        """Check if a delegation has been revoked."""
        return delegation_id in self._revoked

    def get(self, delegation_id: str) -> Optional[Revocation]:
        """Get the revocation for a delegation, if any."""
        return self._revoked.get(delegation_id)

    def __len__(self) -> int:
        return len(self._revoked)

File: revocation.go
package pact

import (
	"encoding/json"
	"fmt"
	"sync"
	"time"
)

// Revocation represents a signed statement that a delegation has been revoked.
// The revoker must be the "from" entity of the delegation being revoked.
type Revocation struct {
	// Type is always "revocation".
	Type string `json:"type"`

	// DelegationID is the ID of the delegation being revoked.
	DelegationID string `json:"delegation_id"`

	// RevokedBy is the identity of the entity revoking (must be the delegator).
	RevokedBy string `json:"revoked_by"`

	// RevokedAt is when the revocation was issued.
	RevokedAt string `json:"revoked_at"`

	// Reason is an optional human-readable reason for revocation.
	Reason string `json:"reason,omitempty"`

	// Signature is the revoker's Ed25519 signature over the canonical revocation.
	Signature string `json:"signature"`
}

// NewRevocation creates a signed revocation for a delegation.
// The revoker must be the "from" entity of the delegation and must hold the private key.
func NewRevocation(delegation *Delegation, revoker *Identity, reason string) (*Revocation, error) {
	if revoker.PrivateKey == nil {
		return nil, fmt.Errorf("pact: revoker must have a private key")
	}

	// Verify the revoker is the delegator
	if revoker.ID != delegation.From.ID {
		return nil, fmt.Errorf("pact: only the delegator (%s) can revoke, not %s", delegation.From.ID, revoker.ID)
	}

	rev := &Revocation{
		Type:         "revocation",
		DelegationID: delegation.ID,
		RevokedBy:    revoker.ID,
		RevokedAt:    time.Now().UTC().Format(time.RFC3339),
		Reason:       reason,
		Signature:    "",
	}

	// Sign the revocation
	payload, err := canonicalJSON(rev)
	if err != nil {
		return nil, fmt.Errorf("pact: failed to canonicalize revocation: %w", err)
	}

	sig, err := revoker.Sign(payload)
	if err != nil {
		return nil, err
	}
	rev.Signature = encodeBase64URL(sig)

	return rev, nil
}

// VerifyRevocation checks that a revocation is properly signed by the claimed revoker.
// The caller must provide the revoker's public key (from the delegation's "from" field).
func VerifyRevocation(rev *Revocation, revokerPublicKey string) (bool, error) {
	pubKeyBytes, err := decodeBase64URL(revokerPublicKey)
	if err != nil {
		return false, fmt.Errorf("pact: invalid revoker public key: %w", err)
	}

	// Verify the revoker ID matches the key
	expectedID := deriveID(pubKeyBytes)
	if expectedID != rev.RevokedBy {
		return false, fmt.Errorf("pact: revoker ID does not match public key")
	}

	// Decode signature
	sigBytes, err := decodeBase64URL(rev.Signature)
	if err != nil {
		return false, fmt.Errorf("pact: invalid revocation signature: %w", err)
	}

	// Reconstruct the signed payload
	revCopy := *rev
	revCopy.Signature = ""
	payload, err := canonicalJSON(&revCopy)
	if err != nil {
		return false, fmt.Errorf("pact: failed to canonicalize revocation for verification: %w", err)
	}

	return ed25519Verify(pubKeyBytes, payload, sigBytes), nil
}

// ed25519Verify wraps the verify call (imported via identity.go's package).
func ed25519Verify(pubKey, message, sig []byte) bool {
	id := IdentityFromPublicKey(EntityAgent, pubKey)
	return id.Verify(message, sig)
}

// RevocationStore is an interface for storing and checking revocations.
// Providers implement this to check incoming requests against known revocations.
type RevocationStore interface {
	// Add stores a verified revocation.
	Add(rev *Revocation) error

	// IsRevoked checks if a delegation ID has been revoked.
	IsRevoked(delegationID string) bool

	// List returns all known revocations.
	List() []*Revocation
}

// MemoryRevocationStore is a simple in-memory revocation store.
// Suitable for single-process providers. For distributed systems,
// use a shared store (Redis, database, etc.).
type MemoryRevocationStore struct {
	mu          sync.RWMutex
	revocations map[string]*Revocation
}

// NewMemoryRevocationStore creates a new in-memory revocation store.
func NewMemoryRevocationStore() *MemoryRevocationStore {
	return &MemoryRevocationStore{
		revocations: make(map[string]*Revocation),
	}
}

// Add stores a revocation.
func (s *MemoryRevocationStore) Add(rev *Revocation) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.revocations[rev.DelegationID] = rev
	return nil
}

// IsRevoked checks if a delegation ID has been revoked.
func (s *MemoryRevocationStore) IsRevoked(delegationID string) bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	_, exists := s.revocations[delegationID]
	return exists
}

// List returns all known revocations.
func (s *MemoryRevocationStore) List() []*Revocation {
	s.mu.RLock()
	defer s.mu.RUnlock()
	result := make([]*Revocation, 0, len(s.revocations))
	for _, rev := range s.revocations {
		result = append(result, rev)
	}
	return result
}

// SerializeRevocation serializes a revocation to JSON for publishing.
func SerializeRevocation(rev *Revocation) ([]byte, error) {
	return json.MarshalIndent(rev, "", "  ")
}
